src/python/pants/backend/python/goals/pytest_runner.py

```diff
@@ -378,45 +422,89 @@ async def setup_pytest_for_target(
 class PyTestRequest(TestRequest):
     tool_subsystem = PyTest
     field_set_type = PythonTestFieldSet
+    partitioner_type = PartitionerType.CUSTOM
+
+
+@rule(desc="Partition Pytest", level=LogLevel.DEBUG)
+async def partition_python_tests(
+    request: PyTestRequest.PartitionRequest[PythonTestFieldSet],
+    python_setup: PythonSetup,
+) -> Partitions[PythonTestFieldSet, TestMetadata]:
+    partitions = []
+    compatible_tests = defaultdict(list)
+
+    for field_set in request.field_sets:
+        metadata = TestMetadata(
+            interpreter_constraints=InterpreterConstraints.create_from_compatibility_fields(
+                [field_set.interpreter_constraints], python_setup
+            ),
+            extra_env_vars=tuple(sorted(field_set.extra_env_vars.value or ())),
+            xdist_concurrency=field_set.xdist_concurrency.value,
+            resolve=field_set.resolve.normalized_value(python_setup),
+            compatability_tag=field_set.batch_compatibility_tag.value,
+        )
+
+        if not metadata.compatability_tag:
+            # Tests without a compatibility tag are assumed to be incompatible with all others.
+            partitions.append(Partition((field_set,), metadata))
+        else:
+            # Group tests by their common metadata.
+            compatible_tests[metadata].append(field_set)
+
+    for metadata, field_sets in compatible_tests.items():
+        partitions.append(Partition(tuple(field_sets), metadata))
+
+    return Partitions(partitions)
 
 
 @rule(desc="Run Pytest", level=LogLevel.DEBUG)
-async def run_python_test(
-    batch: PyTestRequest.Batch[PythonTestFieldSet, Any],
+async def run_python_tests(
+    batch: PyTestRequest.Batch[PythonTestFieldSet, TestMetadata],
     test_subsystem: TestSubsystem,
 ) -> TestResult:
-    field_set = batch.single_element
 
-    setup = await Get(TestSetup, TestSetupRequest(field_set, is_debug=False))
+    setup = await Get(
+        TestSetup, TestSetupRequest(batch.elements, batch.partition_metadata, is_debug=False)
+    )
     result = await Get(FallibleProcessResult, Process, setup.process)
 
+    warning_description = batch.elements[0].address.spec
+    if len(batch.elements) > 1:
+        warning_description = (
+            f"batch containing {warning_description} and {len(batch.elements)-1} other files"
+        )
+    if batch.partition_metadata.description:
+        warning_description = f"{warning_description} ({batch.partition_metadata.description})"
```

Comment:
Maybe move this logic into an (inline) method `def warning_description()`, so that it only needs to run if there is a warning?